### ISSUE TITLE
pyright 1.1.398 breaks simple type checking with pandas-stubs

### ISSUE THREAD

**Issue by Dr-Irv** [microsoft/pyright#10248]
**Describe the bug**
If you install `pandas` and `pandas-stubs`, then basic type checking fails with 1.1.398 but works with 1.1.397

**Code or Screenshots**

```python
import pandas as pd
from pandas._typing import TimeUnit
i0 = pd.date_range(start="2022-06-01", periods=10)
s0 = pd.Series(i0)
print(s0.dt.unit)
```

**VS Code extension or command-line**

pyright 1.1.398 fails, pyright 1.1.397 works fine.

Just create an environment and install `pandas` and `pandas-stubs` using `pip`.

Error that is reported is:
```text
tunit.py:2:28 - error: "TimeUnit" is not exported from module "pandas._typing" (reportPrivateImportUsage)
```

**Comment by Dr-Irv** [comment]
Worth noting - this error only occurs if `pandas-stubs` is installed.  If you do this test with a local copy of `pandas-stubs`, then the error does not appear.

**Comment by erictraut** [comment]
This was an intentional change. See [this issue](https://github.com/microsoft/pyright/issues/9008) for details.

When a stub package or a "py.typed" package is installed in your environment, submodules whose names begin with `_` are considered private — for use by other modules within the package but not directly importable by external consumers. The same is true for submodules located within subdirectories whose name begins with `_`.  Does `pandas-stubs` not follow this convention?

For more information about public and private symbols, refer to [this section of the Python typing spec](https://typing.python.org/en/latest/spec/distributing.html#library-interface-public-and-private-symbols).

The `reportPrivateImport` rule in pyright is designed to notify users when they're accessing private symbols exported from a type stub or a "py.typed" package. Such symbols can change over time because they're considered implementation details, so they shouldn't be used by consumers. The rule doesn't apply to local imports or to packages. It also doesn't apply to modules that are not marked "py.typed" because the rules are not well established for "private" vs "public" symbols in this case.

**Comment by Dr-Irv** [comment]
The issue for `pandas` is that `pandas/_typing.py` has typing declarations in the source.  So we are mapping the same (and other) typing declarations in `pandas-stubs/_typing.pyi` .

I'm not sure how to fix this without breaking code out in the wild.

**Comment by Dr-Irv** [comment]
> Hmm. Are the type definitions exported by `pandas/_typing.py` re-exported by any submodules that do not start with an underscore? Do you know why `_typing.py` starts with an underscore? Was that intended to signify something other than "private"?

It's a historical artifact.  Here's what we say in the docs at https://pandas.pydata.org/pandas-docs/stable/development/contributing_codebase.html#pandas-specific-types

"Commonly used types specific to pandas will appear in [pandas._typing](https://github.com/pandas-dev/pandas/blob/main/pandas/_typing.py) and you should use these where applicable. This module is private for now but ultimately this should be exposed to third party libraries who want to implement type checking against pandas."

When it was introduced, we were only using `pandas._typing` to type check the pandas code.  But as things evolved, even though it's not documented, people started importing things from `pandas._typing` .  A quick search on Github for "from pandas._typing" reveals close to 17,000 hits, although many of them might be clones of the pandas repo.

We have an open issue related to this:  https://github.com/pandas-dev/pandas/issues/55231 but it's a lot of work to decide which types become "public" and also document them.  As it is now, the released versions of `pandas` and `pandas-stubs` will not work with pyright 1.1.398, so that is problematic.

One possible way to handle this is that instead of you expanding `reportPrivateImportUsage` to cover this case, maybe have a separate error code to correspond to this, and have that only available at the `strict` level?

**Comment by erictraut** [comment]
It would be hard to justify splitting this out into a separate check just to accommodate a historical artifact (arguably a bug) in one library — even a fairly popular one. This would be creating another hard-to-explain historical artifact that pyright would need to support forever.

Another option is to document the issue so pandas developers know how to work around it while they're waiting for a proper fix in some future release of pandas. They can use `# type: ignore` or `# pyright: ignore[reportPrivateImportUsage]` to suppress the error on a case-by-case basis or they can disable the `reportPrivateImportUsage` check in their config settings.

> the released versions of pandas and pandas-stubs will not work with pyright 1.1.398

"Will not work with" is an overstatement. There are straightforward ways to suppress this particular check for developers who want to do so.

Another option is to revert this extended check in pyright for a while to give the pandas and pandas-stubs maintainers a chance to address this issue. I'd prefer not to do this, but it may be in the best interest of pyright and pandas users. How quickly do you think this could be addressed in the pandas library? Six months? A year?

**Comment by Dr-Irv** [comment]
> Another option is to document the issue so pandas developers know how to work around it while they're waiting for a proper fix in some future release of pandas. They can use `# type: ignore` or `# pyright: ignore[reportPrivateImportUsage]` to suppress the error on a case-by-case basis or they can disable the `reportPrivateImportUsage` check in their config settings.

It's not an issue for the pandas developers (I think - not sure about that).

> "Will not work with" is an overstatement. There are straightforward ways to suppress this particular check for developers who want to do so.

That means the entire community of people using `pandas-stubs` and `pyright` would need to do that.  That's a big ask.

> 
> Another option is to revert this extended check in pyright for a while to give the pandas and pandas-stubs maintainers a chance to address this issue. I'd prefer not to do this, but it may be in the best interest of pyright and pandas users. How quickly do you think this could be addressed in the pandas library? Six months? A year?

I'd have to check with the rest of the `pandas` team.  

Is it possible to place something in `pandas-stubs/_typing.pyi` that would indicate to `pyright` that "code that imports from this file should ignore `reportPrivateImportUsage` " ?  Because that is a fix that can be quickly released.

**Comment by Dr-Irv** [comment]
I want to point out the following:

This error only appears if `pandas-stubs` is installed.  If not (and only `pandas` is installed), then the error doesn't appear, even though someone has code that imports from `pandas._typing`.  Also, if you only have a local copy of `pandas-stubs` and it is not installed, the error does not appear.

So if there is a simple fix to `pandas-stubs` that makes `pyright` not report this error, that would solve the problem.

**Comment by erictraut** [comment]
> Is it possible to place something in `pandas-stubs/_typing.pyi` that would indicate to pyright that "code that imports from this file should ignore reportPrivateImportUsage " ?

Good suggestion. I think there's utility in allowing library developers to indicate that a submodule with an underscore shouldn't be treated as private as it normally would. I'd prefer to come up with a standardized mechanism though, something that's not pyright-specific.

Let me think about this and see if I can come up with a proposal that would make sense for the typing spec. I'm wondering if it would be better to indicate this in `__index__.pyi` rather than `_typing.pyi`. I need to think about it more and play around with some options. Let me know if you have any specific suggestions. If we can come up with something that's consistent with the existing rules in the typing spec, I think we should be able to quickly get consensus on an extension to the spec.

**Comment by Dr-Irv** [comment]
> Let me know if you have any specific suggestions. If we can come up with something that's consistent with the existing rules in the typing spec, I think we should be able to quickly get consensus on an extension to the spec.

With `pyright`, we place `# pyright: strict` in some stubs file to indicate that the `strict` setting is valid for that file.  So something similar for this issue would be fine with me.

So if it would be possible to place something like a `# pyright: private_as_public` at the top of the private file, that would be fine with me.  Not sure what would be the correct thing to do that would work with all type checkers, and we'd certainly be in a situation where if you came up with such an idea, then we'd have code that would work with some type checkers and not others.

**Comment by erictraut** [comment]
I haven't had any time to spend on this yet. I'll let you know when I do.

**Comment by dangotbanned** [comment]
@erictraut huge thank you for all the work you've put into `pyright`!

I agree with (https://github.com/microsoft/pyright/issues/10248#issuecomment-2776370114) and (https://github.com/microsoft/pyright/issues/10248#issuecomment-2776599109).

Just I wanted to provide a non-`pandas` example I've just run into.

### Example
[`pytest.param`](https://github.com/pytest-dev/pytest/blob/9d7bf4e544f6e9d43ab0dd390a820d4dae140a53/src/pytest/__init__.py#L41) is a publicly exported function, but the return type [`ParameterSet`](https://github.com/pytest-dev/pytest/blob/9d7bf4e544f6e9d43ab0dd390a820d4dae140a53/src/_pytest/mark/structures.py#L80-L203) is not exported.

In the [`altair`](https://github.com/vega/altair) test suite, we have a generator [function](https://github.com/vega/altair/blob/a6d0b7d5720cb414f3a3b61e58cdbe2bcbadcce4/tests/__init__.py#L171-L209) that yields on a call to `pytest.param`

#### Minimal Repro

```py
from __future__ import annotations

from typing import TYPE_CHECKING

import pytest

if TYPE_CHECKING:
    from collections.abc import Iterator

    # "ParameterSet" is not exported from module "_pytest.mark" Pylance(reportPrivateImportUsage)
    from _pytest.mark import ParameterSet

def example_generator(*args: object) -> Iterator[ParameterSet]:
    for arg in args:
        yield pytest.param(arg)
```

> [!IMPORTANT]
> We only have a small number of these imports, so a few `# pyright: ignore[reportPrivateImportUsage]`(s) works for *us*.

Another option *might be* defining structural type that reflects the private `pytest` type.
`pyright` wouldn't complain, but changes to the internal type would still cause the same churn that the diagnostic aims to avoid.

### UX Challenges
If I remove the annotation, I can let `pyright` infer the type and avoid the import.

![Image](https://github.com/user-attachments/assets/db19002d-492c-4de5-871a-5838c348147e)

Using that approach would require documenting *outside of* the type system. 
Depending on user config, it may also require `# (pyright|type): ignore[...]`(s)

There's also an interaction with `pylance`'s *Double-click to insert* feature.

> [!NOTE]
> I know that's out of your control, just seems to be working *against* the rule 🤔 

![Image](https://github.com/user-attachments/assets/7b2bf92e-e282-4e49-9f50-44823aadf013)

The import path chosen was *more precise*, but it side-stepped an export available in [`_pytest.mark.__init__`](https://github.com/pytest-dev/pytest/blob/9d7bf4e544f6e9d43ab0dd390a820d4dae140a53/src/_pytest/mark/__init__.py#L43).

![Image](https://github.com/user-attachments/assets/57c9b08d-7475-4a70-9dc2-f61db0f3a16f)

**Comment by erictraut** [comment]
I appreciate that this change to `pandas` will require some work. One way to ease the transition is to maintain the official definitions of these types in `_typing.pyi` for now but re-export them from a public module (e.g. `typing.pyi`). This would eliminate the need to maintain two copies of the types and keep them in sync. This approach (defining types in a private submodule but then re-exporting them from a public module) is commonly used in other libraries. In the future, you could choose to reverse this and move the official definitions into the public module but re-export them from `_typing` to retain backward compatibility. At that stage, you could also consider splitting the definitions (maintaining two copies for a time) and using `@deprecated` to mark the ones that are slated to go away eventually.

Thanks for the two suggestions. I could maybe justify removing `reportPrivateImportUsage` from `basic` type checking, but I couldn't justify removing it from `standard`. The `standard` type checking include all rules that are informed by the typing spec, and `reportPrivateImportUsage` falls under this umbrella. Also, changing the rule defaults in pyright feels a bit like "moving the peas around on the plate" to mask a problem rather than working toward a long-term solution, so this doesn't feel like a great option to me.

I'm not sure I'm following your idea about `py.typed`. Perhaps you could elaborate? Currently, `py.typed` files must either be empty or contain the line `partial`. There's no other structure or format defined for the file. Are you proposing that we define a file format and schema for it (presumably adopting an existing syntax like INI, JSON, or TOML), and then adding the ability within this schema to specify which submodules are public versus private? I proposed something similar to this a while ago (I'm struggling to find the thread at the moment), and it was pretty soundly rejected — mainly because there was a desire not to add more complexity to `py.typed` parsing. Maybe I just did a poor job of selling the idea at the time.

If I were to revert the change for #9008, I'd want to first see a commitment from the `pandas` and `pandas-stubs` maintainers that they will work to address this in a timely manner — preferably with an outline of a solution and a proposed timeline for a fix. If there is no commitment to fix this in `pandas`, then I'm less inclined to revert the change. Do you think that such a commitment is possible?

**Comment by Dr-Irv** [comment]
> I'm not sure I'm following your idea about `py.typed`. Perhaps you could elaborate? Currently, `py.typed` files must either be empty or contain the line `partial`. There's no other structure or format defined for the file. Are you proposing that we define a file format and schema for it (presumably adopting an existing syntax like INI, JSON, or TOML), and then adding the ability within this schema to specify which submodules are public versus private? I proposed something similar to this a while ago (I'm struggling to find the thread at the moment), and it was pretty soundly rejected — mainly because there was a desire not to add more complexity to `py.typed` parsing. Maybe I just did a poor job of selling the idea at the time.

Right now, the spec at PEP 561 (https://peps.python.org/pep-0561/) says that `py.typed` files are not needed for stubs-only packages.  So my suggestion is that if the `py.typed` file is supplied in a stubs package, there is a syntax to specify private modules that can be treated as public.  This syntax would only apply to stubs-only packages.  Or if the conflict in the name of the file is a concern, a file named `private.typed` (or something like that) could be supplied in a stubs-only package that has a list of modules that are to be considered "public" even though they start with an underscore.

> 
> If I were to revert the change for [#9008](https://github.com/microsoft/pyright/issues/9008), I'd want to first see a commitment from the `pandas` and `pandas-stubs` maintainers that they will work to address this in a timely manner — preferably with an outline of a solution and a proposed timeline for a fix. If there is no commitment to fix this in `pandas`, then I'm less inclined to revert the change. Do you think that such a commitment is possible?

As both a `pandas-stubs` maintainer and on the pandas steering committee, I'm pretty confident we can get this addressed within a year.  This is being discussed in https://github.com/pandas-dev/pandas/issues/55231

**Comment by erictraut** [comment]
I think it's unlikely that we could get consensus on creating a different variant of `py.typed` that is only applicable to stub packages. 

It looks like you're making good progress toward reaching a consensus on how to fix this longer term in pandas. Thanks for driving this. I'm going to go ahead and revert the change to pyright. This will be included in the next release.

**Comment by Dr-Irv** [comment]
> It looks like you're making good progress toward reaching a consensus on how to fix this longer term in pandas. Thanks for driving this. I'm going to go ahead and revert the change to pyright. This will be included in the next release.

Much obliged.  Thanks for considering our predicament!

**Comment by Dr-Irv** [comment]
@erictraut I'm working on things to change in `pandas` so we can support this feature in the future.

One thing that I tried to do that didn't work with version 1.1.399 is to do something like this:
User has a source file called `myconfig.py` .
In that source file, the user does the following
```python
from pandas._typing import SomeType as SomeType  # pyright: ignore[reportPrivateImportUsage]
```
In another source file called `mytest.py`, the user does the following:
```python
from myconfig import SomeType
```
Using version 1.1.399, `pandas` would report an error on `mytest.py` saying that it was importing a private import.  I would have thought that the `ignore` statement in `myconfig.py` would then propagate up to import in `mytest.py` .  So we can plan for the future, is that expected?

**Comment by erictraut** [comment]
The "redundant import form" (i.e. `from x import y as y`) applies only to stubs and "py.typed" libraries. It has no meaning (and is therefore not needed) in user source files or in libraries that omit a "py.typed" marker. In those cases, pyright assumes that all symbols imported from other modules into the global namespace are re-exported from the module. 

With that in mind, you can modify the code for `myconfig.py` to `from pandas._typing import SomeType  # pyright: ignore[reportPrivateImportUsage]`, and it should work fine. If 1.1.399 is reporting an error in `mytest.py`, I'd consider that a bug.

One additional note for clarification: A `#type: ignore` comment never "propagates", nor does it affect analysis or diagnostic checks in any way other than to simply suppress a diagnostic on the line on which it occurs.

**Comment by Dr-Irv** [comment]
> With that in mind, you can modify the code for `myconfig.py` to `from pandas._typing import SomeType # pyright: ignore[reportPrivateImportUsage]`, and it should work fine. If 1.1.399 is reporting an error in `mytest.py`, I'd consider that a bug.

I tried this with `pandas-stubs` installed using 1.1.399:
File `myconfig.py` :
```python
from pandas._testing import ensure_clean as eclean  # pyright: ignore[reportPrivateImportUsage]
```

File `mytest.py`: 
```python
import pandas as pd

from myconfig import eclean

DF = pd.DataFrame({"a": [1, 2, 3], "b": [0.0, 0.0, 0.0]})

with eclean() as path:
    DF.to_json(path)
```

With 1.1.399, I get the error:
```text
mytest.py:3:22 - error: "eclean" is not exported from module "pandas._testing" (reportPrivateImportUsage)
```

Not that you'd fix this now, since you reverted the change in 1.1.400, but something to note when you bring back that feature.
